### PR TITLE
sql: fix the handling of current_schema and search_path

### LINKED ISSUE #24472: cockroachdb 2.0 breaks node js driver error: invalid target name: "migrations"
**BUG REPORT**

For cockroachdb 2.0.0 the node driver fails to create a table for unknown reasons:

```
[ERROR] error: invalid target name: "migrations"
    at Connection.parseE (/home/node/node_modules/pg/lib/connection.js:545:11)
    at Connection.parseMessage (/home/node/node_modules/pg/lib/connection.js:370:19)
    at Socket.<anonymous> (/home/node/node_modules/pg/lib/connection.js:113:22)
    at emitOne (events.js:116:13)
    at Socket.emit (events.js:211:7)
    at addChunk (_stream_readable.js:263:12)
    at readableAddChunk (_stream_readable.js:250:11)
    at Socket.Readable.push (_stream_readable.js:208:10)
    at TCP.onread (net.js:607:20)
```

If you execute the same query by hand, it does work for unknown reasons:

```
CREATE TABLE IF NOT EXISTS "migrations" ("id"   SERIAL PRIMARY KEY NOT NULL, "name" VARCHAR (255) NOT NULL, "run_on" TIMESTAMP  NOT NULL) 
```

### PR DESCRIPTION
**Pull request by knz**
Fixes #24472.

Prior to this patch, CockroachDB was following the pg docs to the
letter regarding the relationship between `search_path` and the
"current schema": the current schema was taken to always be the first
item in `search_path`, as indicated in
https://www.postgresql.org/docs/9.6/static/functions-info.html
("current_schema returns the name of the schema that is first in the
search path (or a null value if the search path is empty).")

This in turn impacts the creation of objects (tables, views, etc) when
the name is unqualified: pg also specifies that such objects are to be
created in the current schema.

Unfortunately the pg doc is incorrect, as pg has a more nuanced
behavior: only the _valid_ schemas in `search_path` are considered to
determine the current schema in that way.

This was causing an unfortunate visible compatibility bug: if the user
specified an invalid schema in the search path, subsequent create
operations would fail with CockroachDB and succeed with pg.

This patch corrects this issue by adopting pg's behavior.

Ditto for `current_schemas`.

Release note (sql change, bug fix): the built-in functions
`current_schema` and `current_schemas` now only consider valid
schemas, like PostgreSQL does.

### CODE REVIEW

File: pkg/sql/sem/builtins/builtins.go [hunk 1]
```diff
@@ -2196,6 +2196,7 @@ CockroachDB supports the following flags:
 
 	// Metadata functions.
 
+	// https://www.postgresql.org/docs/9.6/static/functions-info.html
```
Comment (justinj) [inline pkg/sql/sem/builtins/builtins.go]: nit: maybe better to reference the most recent docs?

File: pkg/sql/sem/builtins/builtins.go [hunk 2]
```diff
@@ -2208,6 +2209,7 @@ CockroachDB supports the following flags:
 		},
 	},
 
+	// https://www.postgresql.org/docs/9.6/static/functions-info.html
```
Comment (justinj) [inline pkg/sql/sem/builtins/builtins.go]: ditto

File: pkg/sql/sem/builtins/builtins.go [hunk 3]
```diff
@@ -2223,52 +2225,77 @@ CockroachDB supports the following flags:
...
```
Comment (justinj) [inline pkg/sql/sem/builtins/builtins.go]: ditto

File: pkg/sql/sem/builtins/builtins.go [hunk 4]
```diff
@@ -2223,52 +2225,77 @@ CockroachDB supports the following flags:
...
-				"compatibility with PostgreSQL. For a new CockroachDB application, " +
-				"consider using current_database() instead.",
+			Info: "Returns the current schema.",
 		},
 	},
 
-	// For now, schemas are the same as databases. So, current_schemas
-	// returns the current database (if one has been set by the user)
-	// and the session's database search path.
+	// https://www.postgresql.org/docs/9.6/static/functions-info.html
```
Comment (justinj) [inline pkg/sql/sem/builtins/builtins.go]: ditto

File: pkg/sql/sem/builtins/builtins.go [hunk 5]
```diff
@@ -2223,52 +2225,77 @@ CockroachDB supports the following flags:
...
 					}
 				}
 				return schemas, nil
 			},
-			Info: "Returns the current search path for unqualified names.",
+			Info: "Returns the valid schemas in the search path.",
 		},
 	},
 
+	// https://www.postgresql.org/docs/9.6/static/functions-info.html
```
Comment (justinj) [inline pkg/sql/sem/builtins/builtins.go]: ditto

**Review by justinj — APPROVED** [review]
there's also a `search_path_test.go`, not sure if there's an appropriate way to add this to that, but it does exist. either way LGTM

File: pkg/sql/sem/builtins/builtins.go (hunk 1 above)
Comment (knz) [reply pkg/sql/sem/builtins/builtins.go]: done here and below

**Comment by knz** [comment]
The search_path_test.go is at a lower level than this patch (and does not have access to a lookup context that can check the validity of schemas), so that doesn't seem the right place.

**Pull request merged**